### PR TITLE
feat: implement dynamic pruning strategy

### PR DESCRIPTION
**Pull request by morph-dev**
### What was wrong?

Current pruning logic for `IdIndexedV1Store` is to delete up to 5% of storage, but no more than 100 elements. However, even 100 elements is too slow on big databases (e.g. 35GB that are currently used). It can take up to a second and sometimes even longer. This wouldn't be such a big problem if this wouldn't block other read/writes.

### How was it fixed?

The idea is to have dynamic number of items that we delete. The `PruningConfig` allows us to set parameters, for this dynamic strategy:
- what is the optimal pruning duration - default 100-300 milliseconds
- by how much to increase/decrease number of items to prune when pruning is faster/slower than optimal - default 20%

Also added logic for handling the zero storage capacity to prune the db (if non-empty) and set radius to zero.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Clean up commit history and use [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).

### CODE REVIEW

**Comment by njgheorghita** [comment]
It seems like this might be relevant... Taking into account #1228 and that (IMO) we will sooner rather than later want to support this, along with a "max db" size (aka a user has 100 gbs of disk space they want to contribute, we'll actually spin up 10 nodes of 10gbs under the hood). Once that's implemented, we'll have a consistent max db size, which means we really don't have to worry about improving the pruning mechanism to support large dbs and can optimize db performance against a fixed target (eg a 10gb storage). Assuming that my description is the direction we're heading, does this pr still make sense?

**Comment by pipermerriam** [comment]
> we'll actually spin up 10 nodes of 10gbs under the hood)

I don't think this is how the architecture will actually work out.  I think that under the hood it will be a single 100gb database that is represented in the network across 10 different node-ids.

See: https://github.com/ethereum/portal-network-specs/issues/283#issuecomment-2037869161

File: trin-storage/src/versioned/id_indexed_v1/store.rs
Comment (carver) [inline trin-storage/src/versioned/id_indexed_v1/store.rs:478]: Seems like a bummer to have prune run its own `Instant` timer, when we're already calculating the time in the metrics. I suppose if `stop_process_timer` returned the duration it observed (looks like it's an f64 in seconds, but we could convert to a `Duration`), then `observe_duration` could reuse it. Do you agree that would be preferable? If you don't feel like doing that, I might do it after this merges, just so it's ready the next time we want to re-use the timing.

File: trin-storage/src/versioned/id_indexed_v1/pruning_strategy.rs
Comment (carver) [inline trin-storage/src/versioned/id_indexed_v1/pruning_strategy.rs:35]: Nit: I suppose 1.0 is a valid choice here, which Range seems to exclude. Due to database compression, there was even a time when >1.0 choices might be valid. I haven't observed how SQL performs since the switchover. Anyway it all seems mostly moot since it's hard-coded for now, but just wanted to make a note that I think it would be fine to limit the input up to 2.0 or something, since it would just be about letting people play with configuration values eventually, I don't think it's a massive footgun that we need to panic to avoid letting them play with larger values.

File: trin-storage/src/versioned/id_indexed_v1/pruning_strategy.rs
Comment (carver) [inline trin-storage/src/versioned/id_indexed_v1/pruning_strategy.rs:41]: Similarly, it's not totally obvious to me that 1 is a hard upper limit here. If someone wanted to play with 2x the pruning count change, my intuition isn't to panic to prevent them from playing with it.

_Edit: I see where the number came from now, but stand by the above comment when paired with the suggestion below._

File: trin-storage/src/versioned/id_indexed_v1/pruning_strategy.rs
Comment (carver) [inline trin-storage/src/versioned/id_indexed_v1/pruning_strategy.rs:154]: Ah, I see now where the upper limit of 1 came from. My intuition before getting to this code is that a change fraction of 1.0 would double on the upside and cut in half on the downside. Those seem like the more symmetric options to me. (ie~ one increase and one decrease takes you back to where you started)

In this approach, choosing a value of .9 nearly doubles on the upside, but is 1/10th on the downside, which is aggressively dropping but not that aggressively climbing (it takes more than 3 upside adjustments to make up for 1 downside adjustment). So if we want one number to govern both directions, my intuition says that this line should be:
```suggestion
            1. / (1. + pruning_config.max_pruning_count_change_fraction)
```

**Review by carver — APPROVED** [review]
LGTM! Mostly just nit comments that you're welcome to ignore.

I apparently have a strong opinion about the ideal way to calculate the percentage update of the pruning number (more below). But in the end, it's not a blocker because even if the number isn't ideal, I don't foresee any catastrophe. So don't let me stop you from merging it if you disagree about the percentage update 😆

**Review by ogenev — APPROVED** [review]
⛵

**Review by njgheorghita — APPROVED** [review]

File: trin-storage/src/versioned/id_indexed_v1/pruning_strategy.rs
Comment (morph-dev) [reply trin-storage/src/versioned/id_indexed_v1/pruning_strategy.rs]: Definitely while coding I had a mental model that `target_capacity` should we lower than `storage_capacity`.

Maybe code would work even if that's not the case, but I would say it doesn't makes much sense, because both storage capacity and target capacity represent the raw content size (`content_id.len() + content_key.len() + content_value.len()`), or at least approximation (there is some extra storage used for indexes and other smaller columns). Irrelevant, but I don't think that there is any compression happening at the moment.

With that being said, I will allow value 1.0. And once we have real use case for using values higher than 1, we can come back to it (in which case we would probably have to rewrite some of the logic).

File: trin-storage/src/versioned/id_indexed_v1/pruning_strategy.rs
Comment (morph-dev) [reply trin-storage/src/versioned/id_indexed_v1/pruning_strategy.rs:45]: I actually wanted increasing the `max_pruning_count` to be harder than decreasing, because downside of being too low is not so bad (we prune less items more frequently) while downside of being too high is that db is blocked for longer.

I added better explanation to the `max_pruning_count_change_fraction` field.

However, if the behavior that you expected is more common way of doing it, I'm willing to change it in the followup PR.

**Review by morph-dev — COMMENTED** [review]
@carver I do disagree with the proposed changes about pruning number changes, but not strongly :). Willing to discuss it more before merging.

**Comment by morph-dev** [comment]
> @carver I do disagree with the proposed changes about pruning number changes, but not strongly :). Willing to discuss it more before merging.

I'm going to merge this make a followup PR if needed.

**Pull request merged**
